### DISCUSSION TITLE
Static Output Render Cycle

### DISCUSSION THREAD

**Discussion by kudla** [vercel/next.js#72054] [Help]
### Summary

Hi!
Having such a component
```ts
const RenderCycle = () => {
   const [mainPortal, mountMainPortal] = useState<HTMLDivElement | null>(null)

   return (
      <>
          <div ref={mountMainPortal} />
          <p>
                  portal mounted: { JSON.stringify(!!mainPortal) }
          </p>
         <NestedComponent portal={mainPortal} />
      </>
   )
}
```
The crucial point here a portal ref that is passed to downstream components to render into

While building the application into a static `export` as I understand it runs only a first single render pass so that the `mainPortal ` is not still mounted. Though the portal doesn't have any expected artifact rendered into at the `out/**/*.html` assets.

So my question is how to force a build to be produced only right after all the render "effects" are resolved (I.e. the tree its totally converged into a state of rest) .

While my PRIMARY interest here in this example is exactly on portals/state mutation. Still the question maybe could be applied for any other effect e.g. (timers, etc.) . So could the application be rendered (without  external UI user interaction) only after the state "calm down"?

As I'm quite new into SSR I hope this should be a pretty common problem and there's a ready straightforward solution for that.

Thank you

**Comment by aniruddhaadak80** [comment]
Hey @kudla ,

To ensure `mainPortal` is fully set before rendering `NestedComponent` in a static export build, use conditional rendering with `useLayoutEffect`:

```jsx
import { useState, useLayoutEffect } from 'react';

const RenderCycle = () => {
   const [mainPortal, setMainPortal] = useState(null);

   useLayoutEffect(() => {
      if (mainPortal) {
         // Actions dependent on mainPortal being available
      }
   }, [mainPortal]);

   return (
      <>
         <div ref={setMainPortal} />
         <p>portal mounted: {JSON.stringify(!!mainPortal)}</p>
         {mainPortal && <NestedComponent portal={mainPortal} />}
      </>
   );
};
```

This ensures `NestedComponent` only renders after `mainPortal` is mounted.

Thank you .

**Comment by icyJoseph** [comment]
This is tricky, SSR doesn't wait for effects and such to flush, even `useLayoutEffect` is stressed to the devs as a noop in the server, since it doesn't run before yielding to the browser, since there's none... I am not sure there's a good solution here, other than reworking your HTML structure. Of course, given that your requirement is that, you must include all of these elements, in the SSR HTML string.

**Comment by kudla** [comment]
Thank you guys
@aniruddhaadak80 I omitted some logic in my example for the simplicity. Including conditional rendering.

It was not about safe rendering into an empty ref it self. But about freezing entire DOM tree for Hydration just right after the initial render pass. So that DOM snapshot stored into a static content corresponds to a null ref state. I had some hope that there is some trick around `useLayoutEffect` from your example. Still I've tied and it didn't work. Or maybe I didn't really get your point right.

But looks like @icyJoseph states all this to be sad but true. Actually it is really hard for me to imagine something that would be a proper work around for portals approach which looks almost like a huck it self by design :) 


Anyway thanks for being involved
